enhancements/machine-config/admin-defined-node-disruption-policy.md

```diff
@@ -0,0 +1,311 @@
+---
+title: admin-defined-node-disruption-policy
+authors:
+  - "@yuqi-zhang"
+reviewers:
+  - "@cgwalters"
+approvers:
+  - "@djoshy"
+api-approvers:
+  - "@JoelSpeed"
+creation-date: 2024-02-08
+last-updated: 2024-02-08
+tracking-link:
+  - https://issues.redhat.com/browse/RFE-4079
+see-also:
+  - https://issues.redhat.com/browse/OCPSTRAT-380
+  - https://issues.redhat.com/browse/MCO-507
+replaces:
+superseded-by:
+---
+
+# Admin Defined Node Disruption Policy
+
+## Summary
+
+This enhancement outlines an API/mechanism for users to define what actions to take upon a MachineConfigOperator driven change (e.g. a file change via a MachineConfig object). By default, all changes to MachineConfig fields require a drain and reboot. The user can use this new API, a Node Disruption Policy subfield in the MachineConfiguration object, to specify which MachineConfig changes to not disrupt their workloads at a cluster scope.
+
+## Motivation
+
+Historically, workload disruption (drains and reboots) can be very costly for OCP users, especially in bare-metal environments where the stopping and restarting of a node can take up to an hour. For small incremental changes to the host OS, many of them can take effect with a service reload or no action at all, and thus the ability to perform a non-disruptive update can be very beneficial to many users without any downside.                  
+
+The MCO since OCP 4.7 has added the ability to perform disruptionless updates on select changes, see: [MCO doc on disruptionless updates](https://github.com/openshift/machine-config-operator/blob/master/docs/MachineConfigDaemon.md#rebootless-updates). This list however is by no means exhaustive, and it is not feasible for the MCO to hard-code all use cases for all users into the MCO. In addition, different users may have different expectations on what will cause a disruption (drains and reboots) or not. Thus there is a need for a user-defined, MCO operated list of "node disruption policies" on what to do after an OS update has been applied.
+
+### User Stories
+
+* As an Openshift cluster administrator, I have an environment with thousands of ICSP/IDMS mirror entries. These entries will be added and deleted constantly on a weekly basis. I understand the potential dangers of removing mirrors (can cause an image to no longer be accessible if no other mirror hosts it), but I accept that risk and would like my changes to apply without disrupting my workloads and only by reloading the crio service.
+
+* As an application developer, I have certificates for my applications I need on-disk, which can be rotated safely and hot-loaded into the applications without any additional action. I do not wish to disrupt other applications when these update via the MCO.
+
+### Goals
+
+* Create a NodeDisruptionPolicyConfig subfield in the MachineConfiguration CRD to allow users to define how OS updates are actioned on after application
```

Comment:
ending of this sentence seems a bit confusing. Perhaps we wanted to write something after `application`

Reply:
Edited